mypy/test/testcheck.py

```diff
@@ -107,7 +107,11 @@ def run_case(self, testcase: DataDrivenTestCase) -> None:
             finally:
                 experiments.STRICT_OPTIONAL = False
         else:
-            self.run_case_once(testcase)
+            try:
+                old_strict_optional = experiments.STRICT_OPTIONAL
+                self.run_case_once(testcase)
+            finally:
+                experiments.STRICT_OPTIONAL = old_strict_optional
```

Comment:
It's up to you, but I think this little change could easily be its own PR (and would be trivially mergeable).